Mathlib/Probability/Kernel/Disintegration/Basic.lean

```diff
@@ -0,0 +1,166 @@
+  classical
+  have (a) : ρCond a (Prod.mk a ⁻¹' {x} ×ˢ s) = ({x} : Set α).indicator (ρCond · s) a := by
+    obtain rfl | hax := eq_or_ne a x
+    · simp only [singleton_prod, mem_singleton_iff, indicator_of_mem]
+      congr with y
+      simp
+    · simp only [singleton_prod, mem_singleton_iff, hax, not_false_eq_true, indicator_of_not_mem]
+      have : Prod.mk a ⁻¹' (Prod.mk x '' s) = ∅ := by ext y; simp [Ne.symm hax]
+      simp only [this, measure_empty]
+  simp_rw [this]
+  rw [MeasureTheory.lintegral_indicator _ (measurableSet_singleton x)]
+  simp only [Measure.restrict_singleton, lintegral_smul_measure, lintegral_dirac]
+  rw [← mul_assoc, ENNReal.inv_mul_cancel hx (measure_ne_top _ _), one_mul]
+
+/-- If the singleton `{x}` has non-zero mass for `ρ.fst`, then for all `s : Set Ω`,
+`ρCond x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s)` . -/
+lemma IsCondKernel.apply_of_ne_zero [MeasurableSingletonClass α] {x : α}
+    (hx : ρ.fst {x} ≠ 0) (s : Set Ω) : ρCond x s = (ρ.fst {x})⁻¹ * ρ ({x} ×ˢ s) := by
+  have : ρCond x s = ((ρ.fst {x})⁻¹ • ρ).comap (fun (y : Ω) ↦ (x, y)) s := by
+    congr 2 with s hs
+    simp [IsCondKernel.apply_of_ne_zero_of_measurableSet _ _ hx hs,
+      (measurableEmbedding_prod_mk_left x).comap_apply]
+  simp [this, (measurableEmbedding_prod_mk_left x).comap_apply, hx]
+
+lemma IsCondKernel.isProbabilityMeasure [MeasurableSingletonClass α] {a : α} (ha : ρ.fst {a} ≠ 0) :
+    IsProbabilityMeasure (ρCond a) := by
+  constructor
+  rw [IsCondKernel.apply_of_ne_zero _ _ ha, prod_univ, ← Measure.fst_apply
+    (measurableSet_singleton _), ENNReal.inv_mul_cancel ha (measure_ne_top _ _)]
+
+lemma IsCondKernel.isMarkovKernel [MeasurableSingletonClass α] (hρ : ∀ a, ρ.fst {a} ≠ 0) :
+    IsMarkovKernel ρCond := ⟨fun _ ↦ isProbabilityMeasure _ _ (hρ _)⟩
+
+end MeasureTheory.Measure
+
+/-!
+### Disintegration of kernels
+
+This section provides a predicate for a kernel to disintegrate a kernel. It also proves that if `κ`
+is an s-finite kernel from a countable `α` such that each measure `κ a` is disintegrated by some
+kernel, then `κ` itself is disintegrated by a kernel, namely
+`ProbabilityTheory.Kernel.condKernelCountable`..
+-/
+
+namespace ProbabilityTheory.Kernel
+variable (κ : Kernel α (β × Ω)) (κCond : Kernel (α × β) Ω)
+
+/-! #### Predicate for a kernel to disintegrate a kernel -/
+
+/-- A kernel `κCond` is a conditional kernel for a kernel `κ` if it disintegrates it in the sense
+that `κ.fst ⊗ₖ κCond = κ`. -/
+class IsCondKernel : Prop where
+  disintegrate' : κ.fst ⊗ₖ κCond = κ
+
+instance instIsCondKernel_zero (κCond : Kernel (α × β) Ω) : IsCondKernel 0 κCond where
+  disintegrate' := by simp
+
+variable [κ.IsCondKernel κCond]
+
+lemma disintegrate : κ.fst ⊗ₖ κCond = κ := IsCondKernel.disintegrate'
+
+/-! #### Existence of a disintegrating kernel in a countable space -/
+
+section Countable
+variable [Countable α] (κCond : α → Kernel β Ω)
+  (h_atom : ∀ x y, x ∈ measurableAtom y → κCond x = κCond y)
```

Comment:
I am not a fan of hiding complicated hypotheses in variable lines (for readability reasons). Could you move `h_atom` to each lemma/def?